### PR TITLE
fix array printing when small number of rows

### PR DESCRIPTION
**Pull request by rfourquet**
This is quite similar to #22917 (printing `Dict`). It's rare to have so few rows available on screen, but still worth it to get the printing right (although I expected it takes me less time to fix...)

### CODE REVIEW

**Comment by rfourquet** [comment]
CI failures are one "arnoldi" and one codegen issue... relaunching CI just in case, and then will merge if it appears good.

File: base/show.jl [hunk 1]
```diff
@@ -1578,11 +1578,11 @@ function print_matrix(io::IO, X::AbstractVecOrMat,
                 print(io, i == first(rowsA) ? pre : presp)
                 print_matrix_row(io, X,A,i,colsA,sep)
                 print(io, i == last(rowsA) ? post : postsp)
-                if i != rowsA[end]; println(io); end
+                if i != rowsA[end] || i == rowsA[halfheight]; println(io); end
                 if i == rowsA[halfheight]
                     print(io, i == first(rowsA) ? pre : presp)
                     print_matrix_vdots(io, vdots,A,sep,vmod,1)
-                    println(io, i == last(rowsA) ? post : postsp)
+                    print(io, i == last(rowsA) ? post : postsp*'\n')
```
Comment (Sacha0) [inline base/show.jl]: For readability, perhaps `"($postsp)\n"`, `(postsp * '\n')`, or `string(postsp, '\n')`? (Without spaces the `*'\` construction took a moment to mentally parse.)

File: base/show.jl [hunk 2]
```diff
@@ -1597,16 +1597,22 @@ function print_matrix(io::IO, X::AbstractVecOrMat,
...
                 print(io, i == last(rowsA) ? post : postsp)
-                if i != rowsA[end]; println(io); end
+                if i != rowsA[end] || i == rowsA[halfheight]; println(io); end
                 if i == rowsA[halfheight]
                     print(io, i == first(rowsA) ? pre : presp)
                     print_matrix_vdots(io, vdots,Lalign,sep,vmod,1)
                     print(io, ddots)
                     print_matrix_vdots(io, vdots,Ralign,sep,vmod,r)
-                    println(io, i == last(rowsA) ? post : postsp)
+                    print(io, i == last(rowsA) ? post : postsp*'\n')
```
Comment (Sacha0) [inline base/show.jl]: Likewise here, perhaps tweak for readability?

File: test/show.jl [hunk 3]
```diff
@@ -880,3 +880,28 @@ end
     @test replstr(zeros(Complex{Int}, 1, 2, 1)) ==
         "1×2×1 Array{Complex{$Int},3}:\n[:, :, 1] =\n 0+0im  0+0im"
 end
+
+@testset "Array printing with limited rows" begin
+    buf = IOBuffer()
+    arrstr(A, rows) = (Base.showarray(IOContext(buf, displaysize=(rows, 80), limit=true),
+                                      A, false, header=true);
+                       String(take!(buf)))
```
Comment (Sacha0) [inline test/show.jl]: Perhaps a long-form definition would be better? The indentation is somewhat confusing.

File: test/show.jl [hunk 4]
```diff
@@ -880,3 +880,28 @@ end
...
+    push!(A, 2)
+    @test arrstr(A, 5) == "2-element Array{Int64,1}:\n ⋮"
+    @test arrstr(A, 6) == "2-element Array{Int64,1}:\n 1\n 2"
+    push!(A, 3)
+    @test arrstr(A, 6) == "3-element Array{Int64,1}:\n 1\n ⋮"
+
+    @test arrstr(zeros(4, 3), 4)  == "4×3 Array{Float64,2}: …"
+    @test arrstr(zeros(4, 30), 4) == "4×30 Array{Float64,2}: …"
+    @test arrstr(zeros(4, 3), 5)  =="4×3 Array{Float64,2}:\n ⋮      ⋱  "
+    @test arrstr(zeros(4, 30), 5) =="4×30 Array{Float64,2}:\n ⋮      ⋱  "
```
Comment (Sacha0) [inline test/show.jl]: Missing spaces after the `==` on this and the preceding line?

File: base/show.jl (hunk 1 above)
Comment (rfourquet) [inline base/show.jl]: Ok will add a space.

**Comment by rfourquet** [comment]
Travis failures are timeouts.

**Pull request merged**